### PR TITLE
feat: add rocksdb write_options to configuration

### PR DESCRIPTION
**Pull request by tufitko**
#878
also, I've updated Dockerfile with Prerequisite command, because it didn't build
C++ isn't my common lang, so any comments are welcome

### CODE REVIEW

**Comment by tanruixiang** [comment]
May I ask why change the storage interface. I think it is necessary to reserve this `WriteOptions` parameter. Not all types share the same storage's writeoption(Even though all are now using the same defaults, I think it's still necessary to keep).

**Comment by tufitko** [comment]
@tanruixiang I can keep old Write interface, I've removed it because I didn't see non-default write options. Can you give an example of when this might be needed?

**Comment by tanruixiang** [comment]
> @tanruixiang I can keep old Write interface, I've removed it because I didn't see non-default write options. Can you give an example of when this might be needed?

There are many options in `WriteOptions`, not just `disableWAL`. If someone needs to change other options like you, and use different options in different places, they will need this `WriteOptions` parameter. In particular, the `rocksdb::WriteOptions write_opts_ ` you add are private, which means we can't make changes in places where options might need to be changed.

**Comment by tufitko** [comment]
🤔 @tanruixiang  I think I need to return the old Write interface and add a method to return default write options from storage. Is it ok?

**Comment by mapleFU** [comment]
Hi, I think we can provide a `defaultRocksDBOption` or `writeWithDefault` in `storage`, I think just create default `WriteOptions` everywhere is not a good idea, and only those who know what rocksdb's options means can define them themselves

**Comment by tanruixiang** [comment]
> 🤔 @tanruixiang I think I need to return the old Write interface and add a method to return default write options from storage. Is it ok?

We can discuss it. @git-hulk @PragmaTwice What form do you think is the best?

File: Dockerfile [hunk 1]
```diff
@@ -15,20 +15,20 @@
...
-FROM ubuntu:focal as build
+FROM ubuntu:jammy as build
 
 # workaround tzdata install hanging
 ENV TZ=Asia/Shanghai
 RUN ln -snf /usr/share/zoneinfo/$TZ /etc/localtime && echo $TZ > /etc/timezone
 
 RUN apt update
-RUN apt install -y cmake make git autoconf libtool g++ python3
+RUN apt install -y git gcc g++ make cmake autoconf automake libtool python3 libssl-dev
```
Comment (PragmaTwice) [inline Dockerfile]: I pick this line to #887, since this PR may not be merged quickly. Thanks for your contribution!

**Comment by git-hulk** [comment]
I also prefer to add the `WriteWithDefault` to avoid creating the option everywhere, then we can keep the `Write` method to be consistent with RocksDB.

File: Dockerfile (hunk 1 above)
Comment (tufitko) [reply Dockerfile]: ok! Then I remove it from my PR

**Comment by tufitko** [comment]
So, I've reworked PR with `WriteWithDefault`. What about `Delete`? There was write_options and I removed it.

**Comment by PragmaTwice** [comment]
> So, I've reworked PR with `WriteWithDefault`. What about `Delete`? There was write_options and I removed it.

I think you can recover the WriteOptions parameter, and just expose a method `WriteOptions Storage::DefaultWriteOptions()`.

And you can also add something like `DeleteWithDefault(...) { Delete(DefaultWriteOptions(), ...) }` if you like.

**Comment by tufitko** [comment]
> > So, I've reworked PR with `WriteWithDefault`. What about `Delete`? There was write_options and I removed it.
> 
> I think you can recover the WriteOptions parameter, and just expose a method `WriteOptions Storage::DefaultWriteOptions()`.
> 
> And you can also add something like `DeleteWithDefault(...) { Delete(DefaultWriteOptions(), ...) }` if you like.

reworked again 🙈

File: src/storage.cc [hunk 2]
```diff
@@ -511,7 +512,7 @@ rocksdb::Status Storage::Write(const rocksdb::WriteOptions &options, rocksdb::Wr
     updates->PutLogData(ServerLogData(kReplIdLog, replid_).Encode());
   }
 
-  return db_->Write(options, updates);
+  return db_->Write(write_opts_, updates);
```
Comment (PragmaTwice) [inline src/storage.cc]:
```suggestion
  return db_->Write(options, updates);
```

Comment (tufitko) [reply src/storage.cc]: ooops

Comment (tufitko) [reply src/storage.cc]: fixed

File: src/config.cc [hunk 3]
```diff
@@ -174,6 +174,7 @@ Config::Config() {
       {"rocksdb.delayed_write_rate", false, new Int64Field(&RocksDB.delayed_write_rate, 0, 0, INT64_MAX)},
       {"rocksdb.wal_ttl_seconds", true, new IntField(&RocksDB.WAL_ttl_seconds, 3*3600, 0, INT_MAX)},
       {"rocksdb.wal_size_limit_mb", true, new IntField(&RocksDB.WAL_size_limit_MB, 16384, 0, INT_MAX)},
+      {"rocksdb.disableWAL", true, new YesNoField(&RocksDB.disableWAL, false)},
```
Comment (PragmaTwice) [inline src/config.cc]: I think `disable_wal` is more suitable in configuration file? since other options are all underscore-case and case-insensitive.

**Comment by tanruixiang** [comment]
As @PragmaTwice  said. I think it is possible to increase the configuration of default writeoption in the configuration. Doing this eliminates the need to recompile the code every time the writeoption is modified.

File: src/config.cc (hunk 3 above)
Comment (tufitko) [reply src/config.cc]: `disable_wal` looks better, but in RocksDb it calls `disableWAL`. shouldn't we save the original name?

**Comment by tufitko** [comment]
> As @PragmaTwice said. I think it is possible to increase the configuration of default writeoption in the configuration. Doing this eliminates the need to recompile the code every time the writeoption is modified.

hmmm, I can add `sync`, `no_slowdown`, `low_pri`, `memtable_insert_hint_per_batch`. But `ignore_missing_column_families` and `timestamp` looks useless to setting through the config. do you agree with this?

**Comment by tanruixiang** [comment]
> > As @PragmaTwice said. I think it is possible to increase the configuration of default writeoption in the configuration. Doing this eliminates the need to recompile the code every time the writeoption is modified.
> 
> hmmm, I can add `sync`, `no_slowdown`, `low_pri`, `memtable_insert_hint_per_batch`. But `ignore_missing_column_families` and `timestamp` looks useless to setting through the config. do you agree with this?

I think these parameters can be reserved, and synchronization in the config file(For example, adding an interface to change the `default writeoption`. If the configuration file is configured with `default writeoption`, then make corresponding modifications.). Of course, if it is not in the configuration file, the default value is used.

**Comment by tanruixiang** [comment]
> I think these parameters can be reserved, and synchronization in the config file(For example, adding an interface to change the `default writeoption`. If the configuration file is configured with `default writeoption`, then make corresponding modifications.). Of course, if it is not in the configuration file, the default value is used.

After this modification, if you want to `disableWAL`, you only need to modify the configuration file.

File: src/config.cc (hunk 3 above)
Comment (git-hulk) [reply src/config.cc]: I also found this point, I think it should be inconsistent in RocksDB NOT the intent. `disable_wal` even `disable_WAL` is better than `disableWAL`, what do you think?

**Comment by tufitko** [comment]
@tanruixiang updated, I'm not sure I understood correctly. 
I didn't add `timestamp` because I still think it's useless. also, there isn't it in newer versions of rocksdb.

**Comment by PragmaTwice** [comment]
I prefer to expose only the options that may be used in practice. If we increase the number of options exposed, then we need to consider how kvrocks will behave if these options are used, and whether all aspects of the functionality in kvrocks will still work. 

If we can't be sure of that, then just exposing `disbale_wal` is fine. If you expose other options, you may need to do some additional explaining: such as whether you tested how kvrocks works when those options are used, and under what circumstances they will be used.

BTW, do not forget to update `kvrocks.conf` : )

**Comment by tufitko** [comment]
I need a decision about exposing write options. 
I think `ignore_missing_column_families` and `timestamp` are unneeded for configuring. Others can be useful

**Comment by tanruixiang** [comment]
There are some parameter settings in writeoption that should be safe, such as `sync`. I think it's okay to keep these parameters, and if the user uses a parameter in config file that we're not sure about, then we give a warning (I think for users unfamiliar with writeoption, they'll just use the default values, and users who modify those values are sure Is to have a certain understanding of rocksdb).

**Review by PragmaTwice — DISMISSED** [review]
LGTM

File: src/config.cc [hunk 4]
```diff
@@ -603,6 +611,9 @@ Status Config::finish() {
     auto s = rocksdb::Env::Default()->CreateDirIfMissing(name);
     if (!s.ok()) return Status(Status::NotOK, s.ToString());
   }
+  if (!slaveof_.empty() && RocksDB.write_options.disable_WAL) {
+    return Status(Status::NotOK, "cannot disable the WAL while using Master-Slave replication");
+  }
```
Comment (git-hulk) [inline src/config.cc]: I think we should check this in the `slaveof` command.

**Review by git-hulk — COMMENTED** [review]
Others are good to me except one comment.

File: src/config.cc (hunk 4 above)
Comment (tufitko) [reply src/config.cc]: ok, moved to `Execution` of slaveof

**Review by git-hulk — APPROVED** [review]

File: src/config.cc (hunk 4 above)
Comment (git-hulk) [reply src/config.cc]: Thank you!

**Review by PragmaTwice — APPROVED** [review]

**Pull request merged**